java/ql/src/semmle/code/java/frameworks/spring/SpringWebUtil.qll

```diff
@@ -0,0 +1,241 @@
+        // note for review: the `content` of this wrapper is determined by what's written to it; there's a function `getWriter()`
+        //   that gets the writer used to write to it; should I link these up outside of the csv models?
+        "org.springframework.web.util;ContentCachingResponseWrapper;false;ContentCachingResponseWrapper;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;ContentCachingResponseWrapper;false;getContentAsByteArray;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;ContentCachingResponseWrapper;false;getContentInputStream;;;Argument[-1];ReturnValue;taint",
+        // note for review: adds a cookie (argument 1) to an `HttpServletResponse` (argument 0); whether a cookie should taint a response
+        //   is somewhat unclear to me.
+        "org.springframework.web.util;CookieGenerator;false;addCookie;;;Argument[-1];Argument[0];taint",
+        "org.springframework.web.util;CookieGenerator;false;addCookie;;;Argument[1];Argument[0];taint",
+        "org.springframework.web.util;CookieGenerator;false;getCookieDomain;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;CookieGenerator;false;setCookieDomain;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;CookieGenerator;false;getCookieName;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;CookieGenerator;false;setCookieName;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;CookieGenerator;false;getCookiePath;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;CookieGenerator;false;setCookiePath;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;DefaultUriBuilderFactory;false;DefaultUriBuilderFactory;;;Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;DefaultUriBuilderFactory;false;builder;;;Argument[-1];ReturnValue;taint",
+        "org.springframework.web.util;DefaultUriBuilderFactory;false;expand;;;Argument[-1..0];ReturnValue;taint",
+        "org.springframework.web.util;DefaultUriBuilderFactory;false;expand;(String,Map);;MapValue of Argument[1];ReturnValue;taint",
+        "org.springframework.web.util;DefaultUriBuilderFactory;false;expand;(String,Object[]);;ArrayElement of Argument[1];ReturnValue;taint",
+        "org.springframework.web.util;DefaultUriBuilderFactory;false;getDefaultUriVariables;;;Argument[-1];MapValue of ReturnValue;taint",
+        "org.springframework.web.util;DefaultUriBuilderFactory;false;setDefaultUriVariables;;;MapValue of Argument[0];Argument[-1];taint",
+        "org.springframework.web.util;DefaultUriBuilderFactory;false;uriString;;;Argument[-1];ReturnValue;taint",
+        // note for review: I couldn't find any default xss sanitizer definitions so I've assumed I shouldn't model html escape functions for taint
```

Comment:
That's right. The XSS query is a bit lacking in that regard. Possibly something to look at @atorralba ?

Reply:
Yes, generally I've seen our default sanitizers need some work (the SQLi query is another example of this).

Again, it depends on the type of taint. `htmlEscape` probably helps with sanitizing XSS -- even though [not always](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#why-cant-i-just-html-entity-encode-untrusted-data), it depends on the context of the injection (HTML context vs JavaScript context) -- but it definitely doesn't help with SQLi or RCE, for instance. 

So, probably what I would say is that it indeed propagates taint, but it _also_ is a sanitizer for XSS (for certain XSS sinks, at least).

Reply:
That has been my approach so far -- escaping functions are propagators in more circumstances than they are blockers, unless they radically restrict the charset (e.g., I might not model `toBase64` on the guess that more often than not that's preventing an exploit, and we don't currently have the tagged-flow analysis to require a `fromBase64` on the way to a sink)